scheduler/scd_utils.py

```diff
@@ -0,0 +1,239 @@
+
+        Raises:
+            ValueError: If line parameters are invalid or if line is a duplicate.
+        """
+
+        try:
+            # create datetime from args to see if valid
+            time = dt.datetime.strptime(mmddyy + " " + hhmm, self.scd_dt_fmt)
+        except:
+            raise ValueError("Can not create datetime from supplied formats")
+
+        if not (0 <= prio <= 20):
+            raise ValueError("Priority is out of bounds. 0 <= prio <= 20.")
+
+
+        scd_lines = self.read_scd()
+
+        epoch = dt.datetime.utcfromtimestamp(0)
+        epoch_milliseconds = int((time - epoch).total_seconds() * 1000)
+
+        new_line = {"timestamp" : epoch_milliseconds,
+                    "time" : time,
+                    "duration" : str(duration),
+                    "prio" : str(prio),
+                    "experiment" : experiment}
+
+        if new_line in scd_lines:
+            raise ValueError("Line is a duplicate of an existing line")
+
+        scd_lines.append(new_line)
+
+        # sort first by timestamp, then by duration with default duration first, then by priority.
+        # duration sorting is funky cause the default value is not a int.
+        new_scd = sorted(scd_lines, key=lambda x:(x['timestamp'],
+                                                    (x['duration'] != '-', x['duration']),
+                                                    x['prio']))
+
+        self.write_scd(new_scd)
+
+    def remove_line(self, mmddyy, hhmm, prio, experiment, duration='-'):
+        """Summary
+
+        Args:
+            mmddyy (TYPE): month/day/year string.
+            hhmm (TYPE): hour/minute string.
+            prio (TYPE): priority value.
+            experiment (TYPE): The experiment to run.
+            duration (str, optional): an optional duration to run for.
+
+        Raises:
+            ValueError: If line parameters are invalid or if line does not exist.
+        """
+
+        try:
+            # create datetime from args to see if valid
+            time = dt.datetime.strptime(mmddyy + " " + hhmm, self.scd_dt_fmt)
+        except:
+            raise ValueError("Can not create datetime from supplied formats")
+
+        if not (0 <= prio <= 20):
+            raise ValueError("Priority is out of bounds. 0 <= prio <= 20.")
+
+        scd_lines = self.read_scd()
+
+        epoch = dt.datetime.utcfromtimestamp(0)
+        epoch_milliseconds = int((time - epoch).total_seconds() * 1000)
+
+        line_to_rm = {"timestamp" : epoch_milliseconds,
+                        "time" : time,
+                        "duration" : str(duration),
+                        "prio" : str(prio),
+                        "experiment" : experiment}
+
+        try:
+            scd_lines.remove(line_to_rm)
+        except:
+            raise ValueError("Line does not exist in SCD")
+
+        self.write_scd(scd_lines)
+
+    def get_relevant_lines(self, mmddyy, hhmm):
```

Comment:
This is a great idea and I could see this being handy. I would prefer if it would output the single previous line before the time as well though (so if you give this function a time between two lines, it will output the line that is currently running at the provided time as well as future lines).